### ISSUE TITLE
Glance Decoder Loss

### ISSUE THREAD

**Issue by FraPochetti** [JizhiziLi/GFM#20]
Hi @JizhiziLi , thanks a lot for this awesome work first of all!
I have a question about the GD Loss, please.
In the paper, you mention it is a Cross-Entropy Loss between the output of the GD and the ground truth (dependent on RoSTA).

![image](https://user-images.githubusercontent.com/1206519/112673418-fbf5dd00-8e64-11eb-8826-e596627e5708.png)

According to the code you shared, `glance_sigmoid` is the output of GD and it has shape `(3, H, W)`.
Given you implement the `GFM-TT`, I assume the ground truth here is the one-hot encoded trimap, which will have the same shape.
What confuses me is that you use a sigmoid activation as a very last step of GD.
Shouldn't you use SoftMax? The idea is to classify each pixel in one of 3 classes (foreground, background, unknown).
I am not sure I understand how sigmoid helps here.
Also, pytorch [CELoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) combines LogSoftmax and NLLLoss. Given the input has already gone through sigmoid, it cannot be passed to the CELoss, given LogSoftmax will be applied too.
Shall we "undo" the sigmoid and use CELoss?

Or probably I am overthinking this, and sigmoid is "just" an activation (as ReLU). So I can consider `glance_sigmoid` as the raw output of the decoder and safely pass it to CELoss.

Thanks again!

**Comment by JizhiziLi** [comment]
Hi there,

Sorry for the late reply, we choose sigmoid instead of softmax as the activation function because, in a multi-label classification problem, softmax refers to there is only one "right answer", i.e. the classes are mutually exclusive, while sigmoid refers to the output could belong to two or more classes. In our case, since the definition of transition area is kind of ambiguous, a pixel of the image can be classified into both foreground and transition area or background and transition area at the same time. Thus, we choose sigmoid as our activation function instead of softmax.

But just as you said PyTorch CELoss includes LogSoftmax inside, we should undo the sigmoid and use CELoss directly on the output. We have done the experiment without sigmoid and the results say it will bring a slight improvement in the performance. We will consider updating it in our future version. Thanks!


Cheers,
Jizhizi Li
